samples/excel/85-preview-apis/data-types-custom-functions.yaml

```diff
@@ -0,0 +1,334 @@
+order: 5
+id: excel-data-types-custom-functions
+name: 'Data types: Custom functions'
+description: This sample shows how to write custom functions that return entity value data types.
+host: EXCEL
+api_set:
+    ExcelApi: BETA (PREVIEW ONLY)
+script:
+    content: |
+        /**
+         * Search for products that match a given substring. Try =SCRIPTLAB.DATATYPESCUSTOMFUNCTIONS.PRODUCTSEARCH("syrup").
+         * @customfunction
+         * @param search The string to search for in the sample JSON data.
+         * @param [lookAt] Define whether the search should be a match of the whole product name or part of the product name. If omitted, set to lookAt = "Part".
+         * @return {Promise<any[][]>} Search results as one or more data type entity values.
+         */
+        async function productSearch(query: string, lookAt?: string): Promise<any[][]> {
+          // This method searches a set of sample JSON data for the string entered in the 
+          // custom function, and then returns the search result as one or more entity values.
+
+          // Set up an error to use if a matching product doesn't exist in the JSON data.
+          const notAvailableError = {
+            type: "Error",
+            errorType: "NotAvailable"
+          };
+
+          // Search the sample JSON data for matching product names.
+          try {
+            if (lookAt === null) {
+              lookAt = "Part"
+            }
+            
+            console.log(`Searching for ${query}...`);
+            const searchResult = await searchProduct(query, lookAt);
+
+            // If the search result is empty, return the error.
+            if (searchResult.length == 0) {
+              return [[notAvailableError]];
+            }
+
+            // Create product entities for each of the products returned in the search result.
+            const entities = searchResult.map((product) => [makeProductEntity(product)]);
+            return entities;
+          } catch (error) {
+            console.error(error);
+          }
+        }
+
+        // Helper method to create entities from product properties.
+        function makeProductEntity(product?: any) {
```

Comment:
Are these helper methods recreating the JSON every time? This might be out of the scope of this PR/sample, but if they are, we should consider reworking this to only create the JSON on the first call and reference the stored data every other time (Singleton pattern).

Reply:
Interesting point, definitely open to considering this. @jakobpn Can you answer this question? --> 

> Are these helper methods recreating the JSON every time?

Reply:
Yes, I think that is demonstrating how a real solution that would get the JSON from a service and then construct the entity value for each result. Plus, I think it beyond the purpose of the sample to show how to code a cache mechanism.

Reply:
Ok, thanks! I agree with Jakob and will close this issue. Reopen if you'd like further discussion.